Static list instead of synchronized cache for collection-types

olvesh opened the pull request:
During load-testing of a token service for single sign on I saw a number of wait/block on the 

``` java
   ListType.getInstance(AbstractType)
   SetType.getInstance(AbstractType)
   MapType.getInstance(AbstractType, AbstractType)
```

This pull requests pre-builds maps of the three collection types and thus drops a `synchronized` statement done during each `PreparedStatement.bind(...)`  for each collection-type in that bind.

We have request/responses of ca 3 ms and able to push through 0.8-2.3k req/sec. With this fix I saw a bump to 2.5-3k req/sec.

We are using a 3 node C\* cluster with 32 cores and SSD disks. Plenty fast enough for this. Also we are doing writes and reads at CL1. I saw a near doubling of read/writes on the graph of opscenter, and although it doesn't map up directly with the numbers above, more concurrent traffic is let through.

Hopefully someone else needs this as well.

I am not sure if it is up to your coding-standards - please let me know if something needs to be changed. 

JIRA: https://datastax-oss.atlassian.net/browse/JAVA-222

Commit:
Static list instead of synchronized cache for collection-types

pcmanus commented:
Thanks for the contribution. I forgot those calls where synchronized in Cassandra, that's clearly overkill here.

One detail though is that while nested collections are not allowed, custom types do are allowed inside collections (I doubt many people uses them but still). So we'd need to fallback to getInstance if the static map doesn't return anything. And since we'll only pre-compute the instance for "primitive" types, we can use EnumMaps (with an EnumMap of EnumMap for maps) which ought to be slightly faster.

Regarding code style, I'd drop a 'final' on those static type maps and would probably assign them directly on the line they are declared (instead of having a separate static block) but that's nit-picking. There's also a println in the test file :). Regarding that file, I do would need a license header at the beginning of that test files, and I'd usually rather remove things like the User/Date/Time in the class javadoc.

olvesh commented:
I'll add the check for custom types, with a static method that wraps the map access and returns a call to getInstance via getCodecInternal, but that points to a possible infinite recursion. Do you think it is necesarry to add safeguard in the methods so to check that it is not a nested collection? Or would that case be checked earlier in the pipeline? 
e.g.:

``` java
private static ListType<?> getListType(final DataType type) {
        Preconditions.checkArgument(type.isCollection(), "Only lists can give resolve to a listtype");
        final DataType dataTypeArg0 = type.getTypeArguments().get(0);
        Preconditions.checkArgument(!dataTypeArg0.isCollection(), "No nested collections please!");

        if(dataTypeArg0.getName() == DataType.Name.CUSTOM)
            return ListType.getInstance(getCodecInternal(dataTypeArg0));
        return LISTS.get(type.getTypeArguments().get(0).getName());
    }
```

These preconditions are cheap I guess, but they will be run each time a collection is resolved.
This ok?

pcmanus commented:
>  Do you think it is necesarry to add safeguard in the methods so to check that it is not a nested collection?

No. Nested collections are not allowed by CQL so you are guaranteed to never get those (we'll reconsider if that ever changes, it's not the only place that assumption is made). As for testing the argument of getListType() itself is a collection, we'll only call the method when we know it's a list, so no reason to be over-careful.

olvesh commented:
Also it is impossible (?) to create an EnumMap for the map type as the key is not convertible to a single enum. Would need 256 of them. List and Set will be enum maps though.

Commit:
Switched to using EnumMap for SetType and ListType. Added fallback to 'Collection'Type.getInstance(...) for custom DateType.
Added tests to make sure a custom type is returned and not null (as was the case).

Removed javadoc and added license header for CodecTest.

pcmanus commented:
> Also it is impossible (?) to create an EnumMap for the map type as the key is not convertible to a single enum

Why not using an EnumMap<DataType.Name, EnumMap<DataType.Name, MapType>>? It'll save the Pair allocation every time you want to do a lookup.

olvesh commented:
Would be nice to skip the Pair allocation yes. But as the map-type is 
`Map<Pair<DataType.Name,DataType.Name>, MapType<?,?>>`
how would you build the key? The key here is two enums. Maybe I am missing something obvious though, being late in a long monday and all...

Hm, I guess you are thinking of an `EnumMap<Name, EnumMap<Name, MapType<?,?>>`? 
(it was obvious)

I will see what I get done today, or else tomorrow morning.

Commit:
Swapped usage of Pairs to map of maps

olvesh commented:
There happy now? ;-) 
Should I squash them as well, or is it ok with three commits?

pcmanus commented:
> Should I squash them as well, or is it ok with three commits?

I'll admit I tend to squash before commit myself as I like clean history, but 3 commits is fine and I've bothered you enough :). Commited, thanks again for you contribution. I did committed a few nitpicks related to codestyle, but that's the OCD talking.
